### DISCUSSION TITLE
How to connect to the server through ”wezterm_ssh“ using a password or public key in rust?

### DISCUSSION THREAD

**Discussion by TheBlindM** [wezterm/wezterm#4082] [Q&A] [answered]
I have seen the examples, but I don't quite understand

```
			SessionEvent::Authenticate(auth) => {
				if !auth.username.is_empty() {
					eprintln!("Authentication for {}", auth.username);
				}
				if !auth.instructions.is_empty() {
					eprintln!("{}", auth.instructions);
				}
				let mut terminal = line_editor_terminal()?;
				let mut editor = LineEditor::new(&mut terminal);
				let mut host = PasswordPromptHost::default();
				let mut answers = vec![];
				for prompt in &auth.prompts {
					let mut prompt_lines = prompt.prompt.split('\n').collect::<Vec<_>>();
					editor.set_prompt(prompt_lines.pop().unwrap());
					host.echo = prompt.echo;
					for line in &prompt_lines {
						eprintln!("{}", line);
					}
					if let Some(line) = editor.read_line(&mut host)? {
						answers.push(line);
					} else {
						anyhow::bail!("Authentication was cancelled");
					}
				}
				auth.answer(answers).await?;
			}
			SessionEvent::Error(err) => {
				anyhow::bail!("{}", err);
			}
			SessionEvent::Authenticated => break,
		}
	}
```

I have saved user information in sqlite, including username, host, password, etc. I want to know how to give ssh.

**Comment by TheBlindM** [comment]
Is there anyone? Can you help me?

**Reply by wez** [reply]
The [AuthenticationEvent](https://docs.rs/wezterm-ssh/latest/wezterm_ssh/struct.AuthenticationEvent.html) is delivered to you through the session events channel when interactive authentication (eg: simple password, or other keyboard interactive authentication) is required by the server.

The event contains the series of prompts to which the server expects an answer. The intent of the protocol is that the instructions and prompts are shown to the human user, who then enters their responses.

The `answer` and `try_answer` methods of the `AuthenticationEvent` can be used to provide those responses.

For ssh key authentication, provided that the ssh configuration is set to reference the correct key, and the server accepts it, you don't need to take any additional action in the code, and you will be delivered an `Authenticated` event.

**Reply by TheBlindM** [reply]
I am learning rust and want to implement a management tool for SSH. SSH RS cannot read and write async, so I need to create an SSH session in the code.May I ask if there are any examples?

**Reply by wez** [reply, marked as answer]
both wezterm and https://github.com/chipsenkbeil/distant/ make use of wezterm-ssh

**Reply by TheBlindM** [reply]
Okay, thanku

**Comment by TheBlindM** [comment]
> The [AuthenticationEvent](https://docs.rs/wezterm-ssh/latest/wezterm_ssh/struct.AuthenticationEvent.html) is delivered to you through the session events channel when interactive authentication (eg: simple password, or other keyboard interactive authentication) is required by the server.
> 
> The event contains the series of prompts to which the server expects an answer. The intent of the protocol is that the instructions and prompts are shown to the human user, who then enters their responses.
> 
> The `answer` and `try_answer` methods of the `AuthenticationEvent` can be used to provide those responses.
> 
> For ssh key authentication, provided that the ssh configuration is set to reference the correct key, and the server accepts it, you don't need to take any additional action in the code, and you will be delivered an `Authenticated` event.



> The [AuthenticationEvent](https://docs.rs/wezterm-ssh/latest/wezterm_ssh/struct.AuthenticationEvent.html) is delivered to you through the session events channel when interactive authentication (eg: simple password, or other keyboard interactive authentication) is required by the server.
> 
> The event contains the series of prompts to which the server expects an answer. The intent of the protocol is that the instructions and prompts are shown to the human user, who then enters their responses.
> 
> The `answer` and `try_answer` methods of the `AuthenticationEvent` can be used to provide those responses.
> 
> For ssh key authentication, provided that the ssh configuration is set to reference the correct key, and the server accepts it, you don't need to take any additional action in the code, and you will be delivered an `Authenticated` event.



> The [AuthenticationEvent](https://docs.rs/wezterm-ssh/latest/wezterm_ssh/struct.AuthenticationEvent.html) is delivered to you through the session events channel when interactive authentication (eg: simple password, or other keyboard interactive authentication) is required by the server.
> 
> The event contains the series of prompts to which the server expects an answer. The intent of the protocol is that the instructions and prompts are shown to the human user, who then enters their responses.
> 
> The `answer` and `try_answer` methods of the `AuthenticationEvent` can be used to provide those responses.
> 
> For ssh key authentication, provided that the ssh configuration is set to reference the correct key, and the server accepts it, you don't need to take any additional action in the code, and you will be delivered an `Authenticated` event.



> both wezterm and https://github.com/chipsenkbeil/distant/ make use of wezterm-sshwezterm 和 https://github.com/chipsenkbeil/distant/ 都使用 wezterm-ssh
